vhdx/src/vhdx_header.rs

```diff
@@ -0,0 +1,536 @@
+            ReadingRegionTableEntry(e) => write!(f, "failed to read region table entry: {}", e),
+            ReadingRegionTableHeader(e) => write!(f, "failed to read region table header: {}", e),
+            RegionEntryCollectionFailed => write!(f, "failed to read region entries"),
+            //UnsupportedVersion(v) => write!(f, "unsupported version: {}", v),
+            RegionOverlap => write!(f, "overlapping regions found"),
+            UnrecognizedRegionEntry => write!(f, "we do not recongize this entry"),
+            WritingData(e) => write!(f, "failed to write data: {}", e),
+            WritingHeader(e) => write!(f, "failed to write header: {}", e),
+        }
+    }
+}
+
+#[derive(Clone)]
+struct FileTypeIdentifier {
+    pub signature: u64,
+    pub creator: [u8; 512], //should we keep it here?
+}
+
+const VHDX_SIGN: u64 = 0x656C_6966_7864_6876; //"vhdxfile"
+const FILE_START: u64 = 0; //The first element
+impl FileTypeIdentifier {
+    /// Reads the File Type Identifier structure from a reference VHDX file
+    pub fn new(f: &mut File) -> Result<FileTypeIdentifier> {
+        f.seek(SeekFrom::Start(FILE_START))
+            .map_err(Error::ReadingHeader)?;
+        let signature = f.read_u64::<LittleEndian>().map_err(Error::ReadingHeader)?;
+        if signature != VHDX_SIGN {
+            return Err(Error::InvalidVHDXSign);
+        }
+
+        let mut creator = [0; 512];
+        f.read_exact(&mut creator).map_err(Error::ReadingHeader)?;
+
+        Ok(FileTypeIdentifier { signature, creator })
+    }
+}
+
+#[derive(Clone)]
+struct Header {
+    pub signature: u32,
+    pub checksum: u32,
+    pub sequence_number: u64,
+    pub file_write_guid: u128,
+    pub data_write_guid: u128,
+
+    pub log_guid: u128,
+    pub log_version: u16,
+    pub version: u16,
+    pub log_length: u32,
+    pub log_offset: u64,
+}
+
+const HEADER_SIZE: u64 = 4096; // Each header size if 4KB
+const HEADER_SIGN: u32 = 0x6461_6568; //"head"
+const HEADER1_START: u64 = 64 * 1024; //header 1 start in Bytes
+const HEADER2_START: u64 = 128 * 1024; //header 2 start in Bytes
+
+impl Header {
+    /// Reads the Header structure from a reference VHDX file
+    pub fn new(f: &mut File, start: u64) -> Result<Header> {
+        // Read the whole header in to a buffer. We will need if for calculating checksum
+        let mut buffer = [0; HEADER_SIZE as usize];
+        f.seek(SeekFrom::Start(start))
+            .map_err(Error::ReadingHeader)?;
+        f.read_exact(&mut buffer).map_err(Error::ReadingHeader)?;
+
+        let signature = LittleEndian::read_u32(&buffer[0..3]);
+        if signature != HEADER_SIGN {
+            return Err(Error::InvalidHeaderSign);
+        }
+
+        let checksum = LittleEndian::read_u32(&buffer[4..7]);
+        let new_checksum = match calculate_checksum(&mut buffer, size_of::<u32>()) {
+            Ok(r) => r,
+            Err(e) => panic!("Failed to calculate checksum for Header {}", e),
+        };
```

Comment:
Why are you calling `panic!()` in case of error? Shouldn't you propagate the error instead?
```suggestion
        let new_checksum = calculate_checksum(&mut buffer, size_of::<u32>())
                .map_err(Error::CalculateChecksum)?;
```

Reply:
Done.